063-topic-replication-factor-change.md

```diff
@@ -0,0 +1,240 @@
+
+The `replication_throttle` parameter is an upper bound on the bandwidth used to move replicas (bytes per second).
+This is configurable for all endpoints in `Kafka.spec.cruiseControl.config` with `default.replication.throttle` property.
+
+The `skip_rack_awareness_check` boolean parameter is undocumented and disabled by default.
+In case of a single-rack deployment, rack goals are not configured and the Topic Operator will automatically enable `skip_rack_awareness_check`.
+
+#### User tasks endpoint
+
+The `user_tasks` endpoint is used to periodically check the tasks state by `User-Task-ID`, which is returned in the `topic_configuration` response.
+
+```sh
+$ curl -s "localhost:9090/kafkacruisecontrol/user_tasks?user_task_ids=5344ca89-351f-4565-8d0f-9aade00e053d&json=true" | jq
+{
+  "userTasks": [
+    {
+      "Status": "Completed",
+      "ClientIdentity": "127.0.0.1",
+      "RequestURL": "POST /kafkacruisecontrol/topic_configuration?topic=my-topic&replication_factor=2&skip_rack_awareness_check=true&dryrun=false&json=true",
+      "UserTaskId": "5344ca89-351f-4565-8d0f-9aade00e053d",
+      "StartMs": "1701961135978"
+    }
+  ],
+  "version": 1
+}
+```
+
+### Topic Operator
+
+At runtime, the Topic Operator watches and periodically reconciles the configuration of all managed and unpaused KafkaTopic resources.
+Among the other configurations, the reconciliation logic also detects `.spec.replicas` changes by comparing with the actual topic replication factor in Kafka.
+From the point of view of the Topic Operator, a topic replication factor change can be in the following states:
+
+- **Pending**: replicas is updated, but the replication factor change request is not sent to or accepted by Cruise Control.
+- **Ongoing**: replicas is updated and Cruise Control has accepted the request (task queued), but the execution is not completed (possible task states: Active and InProgress).
+- **Ready**: replicas is updated and Cruise Control task execution is completed (possible task states: Completed).
+
+When Cruise Control integration is enabled and a replicas change is detected by the BatchingTopicController, the event is notified to the ReplicasChangeManager.
+The ReplicasChangeManager maintains a cache of pending and ongoing replicas changes, manages the communication with Cruise Control's REST API and updates the KafkaTopic status.
+
+![Replication Factor Change](./images/063-topic-replication-factor-change.png)
+
+The KafkaTopic resources with replicas change are added to a pendingChanges cache (Java List) the first time they are notified to the ReplicasChangeManager.
+When a replicas change is marked as pending, other replicas changes for the same KafkaTopic are ignored until the pending replicas change is sent to Cruise Control and the execution is completed.
+
+The KafkaTopic resources marked as pending are added to an ongoingChanges cache (Java Map keyed by `User-Task-ID`) when the topic configuration request is accepted by Cruise Control.
```

Comment:
Where do yu get the Task Id when it is not yet given to you by CC?

Reply:
A pending KT are added to the ongoing map only when we have the User-Task-ID (i.e. CC has received the request, queued the task, and replied).